### PR TITLE
Add course reload button and automatic quiz display (visible)

### PR DESCRIPTION
**Pull request by krusche**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.

### Motivation and Context
This is a performance improvement. Some students tend reload the whole Artemis application if they do not see certain exercises or lectures. However, this can lead to performance problems, especially when > 1000 students reload the browser in lectures at the same time.

### Description
This PR adds a for dashboard REST call for a single course and adds a reload button that triggers this call. This should reduce the performance impact on the server significantly compared to a browser reload. In addition, quizzes should be displayed automatically (using web sockets), when they are set to visible

- [x] A quiz which is visible, but not yet started (without start time), should be displayed at the top of the exercise list

### Steps for Testing
1. Log in to Artemis
2. Navigate to Course Overview. Double check that everything is shown correctly (and nothing is missing)
3. Navigate into one course
4. Click the reload button (student) --> nothing should change
5. Add a new exercises or lecture in a separate browser window / tab (instructor)
6. Click on the reload button: the exercise or lecture should be shown now (student)
7. Set a quiz without start time to visible (instructor)
8. The quiz should appear and be shown at the top (student)

### CODE REVIEW

**Comment by stefanwaldhauser** [comment]
image
Even though i am in WaldhauserSCourse it shows an exercise of testCourse

**Comment by alexmardale** [comment]
> image
> Even though i am in WaldhauserSCourse it shows an exercise of testCourse

I just tried this as well, it seems that if you are on a different course than the one the quiz is created for, the quiz will be displayed to you when it is created; but if you refresh with the "Refresh" button or full page refresh it is not displayed anymore

**Comment by stefanwaldhauser** [comment]
Everytime I reload the page the quiz shows up

**Comment by alexmardale** [comment]
For me only when it was created

**Comment by krusche** [comment]
Thanks for finding this bug. I just fixed it, so that the quiz exercise is only sent on the course channel

**Comment by stefanwaldhauser** [comment]
The Reload button is now gone?
image

**Comment by alexmardale** [comment]
Same, gone for me as well

**Comment by stefanwaldhauser** [comment]
image
Some part of the UI load with a delay and the quiz shows up again
image

**Comment by krusche** [comment]
please note that test server 3 is also used for some automated testing (with automatic deployment). I suggest you use test server 1 for testing this feature

**Comment by stefanwaldhauser** [comment]
- Loading Lecture and Exercise with Reload Button works
- Quiz shows up immediately when I set it to visible

File: src/main/webapp/app/overview/courses.component.ts [hunk 1]
```diff
@@ -38,7 +38,7 @@ export class CoursesComponent implements OnInit {
     }
 
     loadAndFilterCourses() {
-        this.courseService.findAll().subscribe(
+        this.courseService.findAllForDashboard().subscribe(
```
Comment (stefanwaldhauser) [inline src/main/webapp/app/overview/courses.component.ts:41]: This change broke something, no courses are displayed anymore 
image

Comment (TobiasPr) [reply src/main/webapp/app/overview/courses.component.ts:41]: same for me

Comment (stefanwaldhauser) [reply src/main/webapp/app/overview/courses.component.ts:41]: This method does not exist on courseService

Comment (madwau) [reply src/main/webapp/app/overview/courses.component.ts:41]: @stefanwaldhauser `courseService` is an instance of `CourseManagementService`.

Comment (stefanwaldhauser) [reply src/main/webapp/app/overview/courses.component.ts:41]: Ah got it thank you :)

Comment (stefanwaldhauser) [reply src/main/webapp/app/overview/courses.component.ts:41]: Where do I see this in typescript @madwau ?

Comment (madwau) [reply src/main/webapp/app/overview/courses.component.ts:41]: You can see it in the constructor where this dependency is injected.

Comment (stefanwaldhauser) [reply src/main/webapp/app/overview/courses.component.ts:41]: haha oh man i see what you mean know :) I thought you meant something with inheritance. The variable name should probably changed to courseManagementService though

Comment (madwau) [reply src/main/webapp/app/overview/courses.component.ts:41]: I'm guessing `courseManagementService` was seen as too verbose, not sure though.

Comment (krusche) [reply src/main/webapp/app/overview/courses.component.ts:41]: I am currently working on a fix. The problem is that a visible quiz has no due date, but this was expected for the sorting of the exercise in the course overview

**Comment by krusche** [comment]
I think I fixed it, please try again (when the build is finished)

**Comment by krusche** [comment]
I also fixed the order so that active quizzes are prioritizes before visible quizzes

**Comment by fde312** [comment]
I have a quiz currently ongoing
image

but on the course overview dashboard it still shows not exercise with the next due date.
image

My understanding is that on the course overview dashboard it should display the same question as in course exercise dashboard.

**Comment by krusche** [comment]
> worked for me. Step 7 / 8 (I had to click the refreseh button, so that the quiz pops up, but I am pretty certain that was intended)
> 
> image
> 
> Is it normal that I see the TestReloadButtonQuiz, twice in the overview?

Yes, the row is always shown twice (also in production). I don't like this, but if at all this should be changed at a later stage.

Steps 7/8 should also work without clicking the refresh button. You need to have 2 browser windows open and ideally two different accounts.

**Comment by krusche** [comment]
> I have a quiz currently ongoing
> image
> 
> but on the course overview dashboard it still shows not exercise with the next due date.
> image
> 
> My understanding is that on the course overview dashboard it should display the same question as in course exercise dashboard.
> 
> Other then that it works good for me.

If you have multiple active (or visible) quizzes in multiple courses, then it shows the first one which is found. If there is really only one active (or visible), then it should be the same

**Comment by TobiasPr** [comment]
I tried again, the quiz does not appear when not using the refresh button, see screenshot. I used different users one instructor, one student
image

after clicking the button it appeared:
image

**Comment by krusche** [comment]
I fixed a small typo with a space in the WS topic subscription, now the quiz should automatically appear for the student when the instructor sets it visible.

The only missing TODO is that the ordering should be retriggered in course-`overview.component.ts:91` so that the visible quiz automatically appears at the top

File: src/main/webapp/app/overview/courses.component.ts [hunk 2]
```diff
@@ -52,22 +54,34 @@ export class CoursesComponent implements OnInit {
...
-                        this.nextRelevantCourse = course;
-                    }
+                    relevantExerciseForCourse.course = course;
+                    relevantExercises.push(relevantExerciseForCourse);
                 }
             });
+            if (relevantExercises.length === 0) {
+                return undefined;
+            } else if (relevantExercises.length === 1) {
+                relevantExercise = relevantExercises[1];
```
[the remark this answers is not in the document]
Comment (krusche) [reply src/main/webapp/app/overview/courses.component.ts:71]: good point, thanks

[the remark this answers is not in the document]
Comment (krusche) [reply src/main/webapp/app/overview/courses.component.ts]: is fixed

**Comment by krusche** [comment]
> I fixed a small typo with a space in the WS topic subscription, now the quiz should automatically appear for the student when the instructor sets it visible.
> 
> The only missing TODO is that the ordering should be retriggered in `course-overview.component.ts:91` so that the visible quiz automatically appears at the top

Ok, I was wrong, this already works as expected. I had 2 visible quizzes, therefore the order did not change. If there are no visible quizzes (i.e. no exercise in the top row) in the exercise overview and the instructor sets one quiz to visible, it should now appear automatically at the top.

Please wait for the latest build on Bamboo, deploy to test server 1 and try it out again.

**Pull request merged**